File: FFRv2/app/src/main/java/com/devingotaswitch/rankings/sources/ParseFFTB.java
package com.devingotaswitch.rankings.sources;

import com.devingotaswitch.rankings.domain.Player;
import com.devingotaswitch.rankings.domain.Rankings;
import com.devingotaswitch.rankings.domain.Team;
import com.devingotaswitch.utils.Constants;
import com.devingotaswitch.utils.GeneralUtils;
import com.devingotaswitch.utils.JsoupUtils;
import com.devingotaswitch.utils.ParsingUtils;

import java.io.IOException;
import java.util.List;

public class ParseFFTB {

    public static void parseFFTBRankingsWrapper(Rankings rankings)
            throws IOException {
        String teams = String.valueOf(rankings.getLeagueSettings().getTeamCount());
        parseFFTBPage(
                rankings,
                "http://www.fftoolbox.com/football/" + Constants.YEAR_KEY + "/auction-values.cfm?pos=QB&teams=" + teams + "&budget=200", Constants.QB);
        parseFFTBPage(
                rankings,
                "http://www.fftoolbox.com/football/" + Constants.YEAR_KEY + "/auction-values.cfm?pos=RB&teams=" + teams + "&budget=200", Constants.RB);
        parseFFTBPage(
                rankings,
                "http://www.fftoolbox.com/football/" + Constants.YEAR_KEY + "/auction-values.cfm?pos=WR&teams=" + teams + "&budget=200", Constants.WR);
        parseFFTBPage(
                rankings,
                "http://www.fftoolbox.com/football/" + Constants.YEAR_KEY + "/auction-values.cfm?pos=TE&teams=" + teams + "&budget=200", Constants.TE);
        parseFFTBPage(
                rankings,
                "http://www.fftoolbox.com/football/" + Constants.YEAR_KEY + "/auction-values.cfm?pos=PK&teams=" + teams + "&budget=200", Constants.K);
        parseFFTBPage(
                rankings,
                "http://www.fftoolbox.com/football/" + Constants.YEAR_KEY + "/auction-values.cfm?pos=Def&teams=" + teams + "&budget=200", Constants.DST);
    }

    private static void parseFFTBPage(Rankings rankings, String url, String pos)
            throws IOException {
        List<String> brokenUp = JsoupUtils.parseURLWithoutUAOrTls(url, "td");
        int min = 0;
        for (int i = 0; i < brokenUp.size(); i++) {
            if (GeneralUtils.isInteger(brokenUp.get(i))) {
                min = i + 1;
                break;
            }
        }
        for (int i = min; i < brokenUp.size(); i += 8) {
            if (i + 8 > brokenUp.size()) {
                break;
            }

            String name;
            String team = ParsingUtils.normalizeTeams(brokenUp.get(i + 1));
            if (Constants.DST.equals(pos)) {
                name = ParsingUtils.normalizeDefenses(team);
            } else {
                name = ParsingUtils.normalizeNames(brokenUp.get(i));
            }
            String age = brokenUp.get(i + 4);
            String exp = brokenUp.get(i + 5);
            String bye = brokenUp.get(i + 3);

            if (team.split(" ").length <= 3) {
                boolean isNewPlayer = false;
                String playerId = name +
                        Constants.PLAYER_ID_DELIMITER +
                        team +
                        Constants.PLAYER_ID_DELIMITER +
                        pos;
                Player player;
                if (rankings.getPlayers().containsKey(playerId)) {
                    player = rankings.getPlayer(playerId);
                } else {
                    // FF Toolbox rankings are ass, so we'll just default to 1 if we haven't seen it yet.
                    // They have retired players for $30+, so their value is suspect.
                    isNewPlayer = true;
                    player = ParsingUtils.getPlayerFromRankings(name,team,pos, 1.0);
                }

                if (GeneralUtils.isInteger(age)) {
                    player.setAge(Integer.parseInt(age));
                }
                if (GeneralUtils.isInteger(exp)) {
                    player.setExperience(Integer.parseInt(exp));
                } else if ("R".equals(exp)) {
                    player.setExperience(0);
                }

                if (isNewPlayer) {
                    rankings.processNewPlayer(player);
                } else {
                    rankings.getPlayers().put(player.getUniqueId(), player);
                }
            }
            Team newTeam = new Team();
            newTeam.setBye(bye);
            newTeam.setName(team);
            rankings.addTeam(newTeam);
        }
    }
}

File: FFRv2/app/src/main/java/com/devingotaswitch/rankings/sources/ParseFFTB.kt
package com.devingotaswitch.rankings.sources

import com.devingotaswitch.rankings.domain.Player
import com.devingotaswitch.rankings.domain.Rankings
import com.devingotaswitch.rankings.domain.Team
import com.devingotaswitch.utils.Constants
import com.devingotaswitch.utils.GeneralUtils.isInteger
import com.devingotaswitch.utils.JsoupUtils.parseURLWithoutUAOrTls
import com.devingotaswitch.utils.ParsingUtils.getPlayerFromRankings
import com.devingotaswitch.utils.ParsingUtils.normalizeDefenses
import com.devingotaswitch.utils.ParsingUtils.normalizeNames
import com.devingotaswitch.utils.ParsingUtils.normalizeTeams
import java.io.IOException

object ParseFFTB {
    @Throws(IOException::class)
    fun parseFFTBRankingsWrapper(rankings: Rankings) {
        val teams = java.lang.String.valueOf(rankings.leagueSettings.teamCount)
        parseFFTBPage(
                rankings,
                "http://www.fftoolbox.com/football/" + Constants.YEAR_KEY + "/auction-values.cfm?pos=QB&teams=" + teams + "&budget=200", Constants.QB)
        parseFFTBPage(
                rankings,
                "http://www.fftoolbox.com/football/" + Constants.YEAR_KEY + "/auction-values.cfm?pos=RB&teams=" + teams + "&budget=200", Constants.RB)
        parseFFTBPage(
                rankings,
                "http://www.fftoolbox.com/football/" + Constants.YEAR_KEY + "/auction-values.cfm?pos=WR&teams=" + teams + "&budget=200", Constants.WR)
        parseFFTBPage(
                rankings,
                "http://www.fftoolbox.com/football/" + Constants.YEAR_KEY + "/auction-values.cfm?pos=TE&teams=" + teams + "&budget=200", Constants.TE)
        parseFFTBPage(
                rankings,
                "http://www.fftoolbox.com/football/" + Constants.YEAR_KEY + "/auction-values.cfm?pos=PK&teams=" + teams + "&budget=200", Constants.K)
        parseFFTBPage(
                rankings,
                "http://www.fftoolbox.com/football/" + Constants.YEAR_KEY + "/auction-values.cfm?pos=Def&teams=" + teams + "&budget=200", Constants.DST)
    }

    @Throws(IOException::class)
    private fun parseFFTBPage(rankings: Rankings, url: String, pos: String) {
        val brokenUp = parseURLWithoutUAOrTls(url, "td")
        var min = 0
        for (i in brokenUp.indices) {
            if (isInteger(brokenUp[i])) {
                min = i + 1
                break
            }
        }
        var i = min
        while (i < brokenUp.size) {
            if (i + 8 > brokenUp.size) {
                break
            }
            var name: String?
            val team = normalizeTeams(brokenUp[i + 1])
            name = if (Constants.DST == pos) {
                normalizeDefenses(team)
            } else {
                normalizeNames(brokenUp[i])
            }
            val age = brokenUp[i + 4]
            val exp = brokenUp[i + 5]
            val bye = brokenUp[i + 3]
            if (team!!.split(" ").size <= 3) {
                var isNewPlayer = false
                val playerId = name +
                        Constants.PLAYER_ID_DELIMITER +
                        team +
                        Constants.PLAYER_ID_DELIMITER +
                        pos
                var player: Player
                if (rankings.players.containsKey(playerId)) {
                    player = rankings.getPlayer(playerId)
                } else {
                    // FF Toolbox rankings are ass, so we'll just default to 1 if we haven't seen it yet.
                    // They have retired players for $30+, so their value is suspect.
                    isNewPlayer = true
                    player = getPlayerFromRankings(name!!, team, pos, 1.0)
                }
                if (isInteger(age)) {
                    player.age = age.toInt()
                }
                if (isInteger(exp)) {
                    player.experience = exp.toInt()
                } else if ("R" == exp) {
                    player.experience = 0
                }
                if (isNewPlayer) {
                    rankings.processNewPlayer(player)
                } else {
                    rankings.players[player.uniqueId] = player
                }
            }
            val newTeam = Team()
            newTeam.bye = bye
            newTeam.name = team
            rankings.addTeam(newTeam)
            i += 8
        }
    }
}